sycl/unittests/kernel-and-program/MultipleDevsCache.cpp

```diff
@@ -0,0 +1,214 @@
+//==--- KPCache.cpp --- KernelProgramCache for multiple devices unit test --==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#define SYCL2020_DISABLE_DEPRECATION_WARNINGS
+
+#include "HelperKernelInfo.hpp"
+#include "detail/context_impl.hpp"
+#include "detail/kernel_bundle_impl.hpp"
+#include "detail/kernel_program_cache.hpp"
+#include <CL/sycl.hpp>
+#include <helpers/CommonRedefinitions.hpp>
+#include <helpers/PiImage.hpp>
+#include <helpers/PiMock.hpp>
+
+#include <gtest/gtest.h>
+
+#include <iostream>
+
+using namespace sycl;
+
+class MultTestKernel {
+public:
+  void operator()(cl::sycl::item<1>){};
+};
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+template <> struct KernelInfo<MultTestKernel> : public MockKernelInfo {
+  static constexpr const char *getName() { return "MultTestKernel"; }
+};
+} // namespace detail
+} // namespace sycl
+} // __SYCL_INLINE_NAMESPACE(cl)
+
+static sycl::unittest::PiImage generateDefaultImage() {
+  using namespace sycl::unittest;
+
+  PiPropertySet PropSet;
+
+  std::vector<unsigned char> Bin{0, 1, 2, 3, 4, 5}; // Random data
+
+  PiArray<PiOffloadEntry> Entries = makeEmptyKernels({"MultTestKernel"});
+
+  PiImage Img{PI_DEVICE_BINARY_TYPE_SPIRV,            // Format
+              __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64, // DeviceTargetSpec
+              "",                                     // Compile options
+              "",                                     // Link options
+              std::move(Bin),
+              std::move(Entries),
+              std::move(PropSet)};
+
+  return Img;
+}
+
+static sycl::unittest::PiImage Img = generateDefaultImage();
+static sycl::unittest::PiImageArray<1> ImgArray{&Img};
+
+static pi_result redefinedContextCreate(
+    const pi_context_properties *properties, pi_uint32 num_devices,
+    const pi_device *devices,
+    void (*pfn_notify)(const char *errinfo, const void *private_info, size_t cb,
+                       void *user_data),
+    void *user_data, pi_context *ret_context) {
+  *ret_context = reinterpret_cast<pi_context>(123);
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedContextRelease(pi_context context) {
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedDevicesGet(pi_platform platform,
+                                     pi_device_type device_type,
+                                     pi_uint32 num_entries, pi_device *devices,
+                                     pi_uint32 *num_devices) {
+  if (num_devices) {
+    *num_devices = static_cast<pi_uint32>(2);
+    return PI_SUCCESS;
+  }
+
+  devices[0] = reinterpret_cast<pi_device>(1111);
+  devices[1] = reinterpret_cast<pi_device>(2222);
```

Comment:
This should only be performed if `devices` is not nil.

Reply:
And not just not nullptr.  If I'm not mistaken, the maximum possible device entries is bound by `num_entries` so the subscripts should check to make sure they are less than num_entries.     Though, in the case of mocks and unit tests, it's less clear to me where this value for `num_entries` comes from and if it is a problem to blindly set `devices[1]`.

Reply:
I added check for nullptr and for num_entries. 
`piDevicesGet` first time is called with `num_entries=0` and not null `num_devices`. Ather the call `num_entries` takes value of `num_devices`. With second call `num_devices` parameter passes with nullptr, so we can create `num_entries` devices. I rewrite `num_devices` to 2, so when `redefinedDevicesGet` is called a second time RT thinks that we have two devices.
This test emulates two devices, so we don't care about real count of devices.